sematic/resolvers/resource_manager.py

```diff
@@ -0,0 +1,91 @@
+# Standard Library
+import abc
+from dataclasses import dataclass, field, replace
+from typing import Dict, List, Tuple
+
+# Sematic
+from sematic.external_resource import ExternalResource
+
+
+class ResourceManager(abc.ABC):
+    """A store for information about external resources and their metadata.
+
+    Abstracts over cloud vs. in-memory implementations. Notably though, this
+    abstraction only covers storage and retrieval of resource metadata--it does
+    not cover activation/deactivation/state updates of the resources themselves.
+    """
+
+    @abc.abstractmethod
+    def get_resource_for_id(self, resource_id: str) -> ExternalResource:
+        pass
+
+    @abc.abstractmethod
+    def save_resource(self, resource: ExternalResource) -> None:
+        pass
+
+    @abc.abstractmethod
+    def link_resource_to_run(self, resource_id: str, run_id: str, root_id: str) -> None:
+        pass
+
+    @abc.abstractmethod
+    def resources_by_root_id(self, root_id: str) -> List[ExternalResource]:
+        pass
+
+
+@dataclass(frozen=True)
+class _InMemoryResourceRecord:
+    """A record of external resources and their associated run(s)
+
+    For now, resources can only be used by one run. However, this might
+    be changed at some point, so we want the database to allow for multiple
+    runs per resource. The in-memory representation for the SilentResolver
+    should mirror the database representation in this regard.
+
+    Attributes
+    ----------
+    resource:
+        The most up-to-date copy of the resource
+    run_id_root_id_pairs:
+        A tuple where each element is a tuple of (<run id>, <root id>)
+    """
+
+    resource: ExternalResource
+    run_id_root_id_pairs: Tuple[Tuple[str, str], ...]
+
+
+@dataclass
+class InMemoryResourceManager(ResourceManager):
```

Comment:
Maybe move to a `resource_managers` directory?

Reply:
done